File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/c/au_332_c_land_i64_i64.c
/*
 * arch-unit 332  --  c  `a && b`  lhs=int64_t rhs=int64_t
 * symbol op_325   outcome LIFTED   4 arch-opcodes
 *
 * the arch-unit, arch-opcode by arch-opcode:
 *   sltu a0, zero, a0                    integer    operator:< unsigned
 *   sltu a1, zero, a1                    integer    operator:< unsigned
 *   c.and a0, a1                         integer    operator:&
 *   c.jr ra                              integer    return
 *
 * answer: int32_t, 32 bits.  parameters are operand bit patterns.
 */
#include "arch_units_int.h"

uint64_t au_332_c_land_i64_i64(uint64_t p0, uint64_t p1)
{
    /* a0: operand `a` (int64_t) arrives in a0 */
    const uint64_t v1 = p0;
    /* a1: operand `b` (int64_t) arrives in a1 */
    const uint64_t v2 = p1;
    const uint64_t v3 = au_sltu(UINT64_C(0x0), v1);
    const uint64_t v4 = au_sltu(UINT64_C(0x0), v2);
    const uint64_t v5 = au_and(v3, v4);
    /* the answer is int32_t, 32 bits */
    return ((v5) & UINT64_C(0xffffffff));
}

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/go/au_332_c_land_i64_i64.go
// arch-unit 332  --  c  `a && b`  lhs=int64_t rhs=int64_t
// symbol op_325   outcome LIFTED   4 arch-opcodes
//
// the arch-unit, arch-opcode by arch-opcode:
//   sltu a0, zero, a0                    integer    operator:< unsigned
//   sltu a1, zero, a1                    integer    operator:< unsigned
//   c.and a0, a1                         integer    operator:&
//   c.jr ra                              integer    return
//
// answer: int32_t, 32 bits.  parameters are operand bit patterns.
package archunits

func Au_332_c_land_i64_i64(p0 uint64, p1 uint64) uint64 {
	// a0: operand `a` (int64_t) arrives in a0
	var v1 uint64 = p0; _ = v1
	// a1: operand `b` (int64_t) arrives in a1
	var v2 uint64 = p1; _ = v2
	var v3 uint64 = au_sltu(uint64(0x0), v1); _ = v3
	var v4 uint64 = au_sltu(uint64(0x0), v2); _ = v4
	var v5 uint64 = au_and(v3, v4); _ = v5
	// the answer is int32_t, 32 bits
	return ((v5) & uint64(0xffffffff))
}
